devicereg/devicereg.proto

```diff
@@ -23,6 +23,44 @@ service DeviceRegistration {
   // the device registration service, which must also delete all streams by
   // calling down to the stream encoder.
   rpc RevokeDevice (RevokeDeviceRequest) returns (RevokeDeviceResponse);
+
+  // CreateEntitlementPolicy is a method exposed by the service which allows a
+  // new entitlement policy to be created and stored within the device
+  // registration service. Once a policy has been created, users will then be
+  // able to apply this policy to their devices via the wallet.
+  rpc CreateEntitlementPolicy (CreateEntitlementPolicyRequest) returns (CreateEntitlementPolicyResponse);
+
+  // DeleteEntitlementPolicy is a method exposed by the service which allows an
+  // authorized client to request that an entitlement policy be deleted.
+  // Deleting a policy will not affect any existing devices that have already
+  // used the policy in order to create one or more streams within the encoder,
+  // however it will prevent any new applications of that policy to other
+  // devices.
+  rpc DeleteEntitlementPolicy (DeleteEntitlementPolicyRequest) returns (DeleteEntitlementPolicyResponse);
+
+  // ListEntitlementPolicies is a method exposed by the service which returns a
+  // list of all policies currently defined and available within the service to
+  // be applied to devices. Currently it just returns a list of all known
+  // policies with no capability to filter or paginate these policies.
+  rpc ListEntitlementPolicies (ListEntitlementPoliciesRequest) returns (ListEntitlementPoliciesResponse);
+
+  // ApplyEntitlement is a method exposed by the service which allows a user who
+  // has previously claimed a device to apply specific entitlement policies to
+  // their device. Applying a policy to a device sets up a new stream on the
+  // stream encoder with the filtering or aggregation functions defined in the
+  // policy applied to incoming events. This will result in the stream encoder
+  // encrypting and emitting the processed events to the datastore for the
+  // specified policy. Note we send the same request message for both this and
+  // RevokeEntitlement as the data sent is exactly the same.
+  rpc ApplyEntitlement (EntitlementRequest) returns (ApplyEntitlementResponse);
```

Comment:
What is the benefit of separating the creation and application of entitlement policies? Is there a case where a policy will be created but not applied?

Reply:
Creating a policy is more like defining a template, e.g. for bcnnow we might define a policy that shares 15 min moving average noise levels only. This template is stored in the device registration service. At this point nothing happens but we do expose this policy to the wallet. When a user decides to apply this policy to their claimed device, they apply it which links a specific device token with the policy template we then call down to the encoder to create a realised instance of the template. This sets up the subscription to mqtt if not already present and adds hooks to process and output to store